daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Damlc.hs

```diff
@@ -428,6 +442,136 @@ execClean projectOpts = do
+        Right LF.DefValue{..}  
+            | ref `Set.member` seen  -> Set.empty
+            | otherwise -> startFromExpr (Set.insert ref seen)  world dvalBody
+        Left _ -> error "This should not happen"
+    LF.EBuiltin _ -> Set.empty
+    LF.ERecCon _ flds -> Set.unions $ map (\(_, exp) -> startFromExpr seen world exp) flds
+    LF.ERecProj _ _ recEx -> startFromExpr seen world recEx
+    LF.ETupleUpd _ recExpr recUpdate -> startFromExpr seen world recExpr `Set.union` startFromExpr seen world recUpdate
+    LF.EVariantCon _ _ varg -> startFromExpr seen world varg
+    LF.ETupleCon tcon -> Set.unions $ map (\(_, exp) -> startFromExpr seen world exp) tcon
+    LF.ETupleProj _ tupExpr -> startFromExpr seen world tupExpr
+    LF.ERecUpd _ _ recExpr recUpdate -> startFromExpr seen world recExpr `Set.union` startFromExpr seen world recUpdate
+    LF.ETmApp tmExpr tmpArg -> startFromExpr seen world tmExpr `Set.union` startFromExpr seen world tmpArg
+    LF.ETyApp tAppExpr _ -> startFromExpr seen world tAppExpr
+    LF.ETmLam _ tmlB -> startFromExpr seen world tmlB
+    LF.ETyLam _ lambdy -> startFromExpr seen world lambdy
+    LF.ECase cas casel -> startFromExpr seen world cas `Set.union` Set.unions ( map ( startFromExpr seen world . LF.altExpr ) casel)
+    LF.ELet (LF.Binding _ e1) e2 -> startFromExpr seen  world e1 `Set.union` startFromExpr seen world e2
+    LF.ENil _ -> Set.empty
+    LF.ECons _ consH consT -> startFromExpr seen world consH `Set.union` startFromExpr seen world consT
+    LF.ESome _ smBdy -> startFromExpr seen world smBdy
+    LF.ENone _ -> Set.empty
+    LF.EUpdate upd -> startFromUpdate seen world upd
+    LF.EScenario _ -> Set.empty
+    LF.ELocation _ e1 -> startFromExpr seen world e1
+    -- x -> Set.unions $ map startFromExpr $ children x
+
+startFromChoice :: LF.World -> LF.TemplateChoice -> DS.Set Action
+startFromChoice world chc = startFromExpr Set.empty world (LF.chcUpdate chc)
+
+templatePossibleUpdates :: LF.World -> LF.Template -> DS.Set Action
+templatePossibleUpdates world tpl = Set.unions $ map (startFromChoice world) (NM.toList (LF.tplChoices tpl))
+
+moduleAndTemplates :: LF.World -> LF.Module -> [(LF.TypeConName ,DS.Set Action)]
+moduleAndTemplates world mod = retTypess
+    where 
+        templates = templatesFromModule mod
+        retTypess = map (\t-> (LF.tplTypeCon t, templatePossibleUpdates world t )) templates
+
+listOfModules :: NM.NameMap LF.Module -> [LF.Module]
+listOfModules modules = NM.toList modules
+
+templatesFromModule :: LF.Module -> [LF.Template]
+templatesFromModule mod = NM.toList $ LF.moduleTemplates mod
+
+dalfsInDar :: Archive -> [BSL.ByteString]
+dalfsInDar dar = [fromEntry e | e <- zEntries dar, ".dalf" `isExtensionOf` eRelativePath e]
+
+dalfBytesToPakage :: BSL.ByteString -> (LF.PackageId, LF.Package)
+dalfBytesToPakage bytes = case Archive.decodeArchive $ BSL.toStrict bytes of
+    Right a -> a
+    Left err -> error (show err)
```

Comment:
I would prefer if we propagate the `Either` to the top/IO instead of turning it into a call to `error`.